%YAML 1.2
# SPDX-FileCopyrightText: NVIDIA CORPORATION & AFFILIATES
# Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0
---
############################################
# This uses the global_pose_tree/pose_tree #
# type_adapter_nitros_context_graph.yaml   #
############################################
name: atlas
components:
- name: bmi088_frame
  type: nvidia::isaac::PoseTreeFrame
  parameters:
    pose_tree: global_pose_tree/pose_tree
---
name: bmi088_driver
components:
  - name: tx_accelerometer
    type: nvidia::gxf::DoubleBufferTransmitter
    parameters:
      capacity: 64
  - name: tx_gyroscope
    type: nvidia::gxf::DoubleBufferTransmitter
    parameters:
      capacity: 64
  - name: async_scheduling_term
    type: nvidia::gxf::AsynchronousSchedulingTerm
  - name: bmi088_driver
    type: nvidia::isaac::Bmi088Driver
    parameters:
      accel_frequency: 200
      gyro_frequency: 200
      accel_index: 0
      gyro_index: 1
      iio_buf_size: 64
      tx_accelerometer: tx_accelerometer
      tx_gyroscope: tx_gyroscope
      imu_frame: atlas/bmi088_frame
      async_scheduling_term: async_scheduling_term
---
name: imu_combiner
components:
  - name: rx_accelerometer
    type: nvidia::gxf::DoubleBufferReceiver
    parameters:
      capacity: 1
      policy: 0
  - name: rx_gyroscope
    type: nvidia::gxf::DoubleBufferReceiver
    parameters:
      capacity: 1
      policy: 0
  - name: tx_imu
    type: nvidia::gxf::DoubleBufferTransmitter
    parameters:
      capacity: 1024
  - name: combiner
    type: nvidia::isaac::ImuMessageCombiner
    parameters:
      rx_accelerometer: rx_accelerometer
      rx_gyroscope: rx_gyroscope
      tx_imu: tx_imu
  - type: nvidia::gxf::MessageAvailableSchedulingTerm
    parameters:
      receiver: rx_accelerometer
      min_size: 1
  - type: nvidia::gxf::MessageAvailableSchedulingTerm
    parameters:
      receiver: rx_gyroscope
      min_size: 1
---
name: sink
components:
- name: signal
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: signal
    min_size: 1
- name: sink
  type: nvidia::isaac_ros::MessageRelay
  parameters:
    source: signal
---
components:
- name: edge0
  type: nvidia::gxf::Connection
  parameters:
    source: bmi088_driver/tx_accelerometer
    target: imu_combiner/rx_accelerometer
- name: edge1
  type: nvidia::gxf::Connection
  parameters:
    source: bmi088_driver/tx_gyroscope
    target: imu_combiner/rx_gyroscope
- name: edge2
  type: nvidia::gxf::Connection
  parameters:
    source: imu_combiner/tx_imu
    target: sink/signal
---
components:
- name: clock
  type: nvidia::gxf::RealtimeClock
- type: nvidia::gxf::GreedyScheduler
  parameters:
    clock: clock
    stop_on_deadlock: false
    check_recession_period_us: 100
- type: nvidia::gxf::JobStatistics
  parameters:
    clock: clock
